doc/Frontend/guide.org

```diff
@@ -0,0 +1,437 @@
+* Syntax Guide
+This document is heavily inspired by the [[http://docs.idris-lang.org/en/latest/reference/syntax-guide.html][Idris syntax guide]].
+** File Organization
+A file contains zero or more [[Top Level Declarations]]
+
+For example
+#+BEGIN_SRC ocaml
+  -- this is a function with a signature!
+  sig foo : Natural
+  let foo = 3
+
+  -- this is a data type declaration
+  type list a = Cons a (List a)
+              | Nil
+
+  -- this is a module declaration!
+  mod Boo =
+    let fi = 10
+  end
+#+END_SRC
+** Comments
+Comments are denoted by two dashes, =--=, and it and all characters up
+until the end of the line are discarded
+
+Example:
+#+BEGIN_SRC ocaml
+  -- This is a comment!
+#+END_SRC
+** Symbols
+Symbols are used for any name declared in the Juvix programming
+language. Symbols are broken into two categories, infix and prefix
+
+Prefix symbols start with either an alpha character or an underscore,
+which can be followed up by any alphanumeric character, underscores,
+punctuation(?, !), or dashes.
+
+#+BEGIN_SRC haskell
+
+  -- a valid symbol
+  hello-there
+
+  -- another valid symbol
+  _unused-symbol
+
+  -- a valid predicate
+  even?
+
+  -- An important action
+  foo!
+
+  -- not a valid prefix symbol
+  -foo
+#+END_SRC
+
+An infix symbol starts with all other symbols including punctuation
+and dashes
+
+#+BEGIN_SRC haskell
+  -- This is a valid infix symbol
+
+  !!
+
+  -- this is also another valid infix symbol
+
+  -<++
+
+  -- this is not a valid infix symbol, but instead a comment of a -
+
+  ---
+
+  -- here is an arrow infix symbol
+
+  ->
+#+END_SRC
+
+** Top Level Declarations
+*** Functions
+
+Functions are started by writing =let= which is followed by any valid
+prefix symbol or an infix symbol surrounded by parenthesis which shall
+be refereed to as the function name. Then, there are zero or more
+arguments, implicit arguments are surrounded in curly braces
+(={}=). The arguments are ended when an equal sign (===) is placed,
+which denotes the start of the body.
+
+Example:
+#+BEGIN_SRC ocaml
+  -- this is a valid function
+
+  let f x = x + 3
+
+  -- this is another valid variable/function
+
+  let y = 5
+
+  -- this is a valid infix symbol
+
+  let (+) = plus
+
+
+  -- a function with an implicit argument
+  let foo {prf} x = x
+#+END_SRC
+
+Another important part of a function is the signature.
+
+A signature is denoted first by sig, then the function name. From here
+colon (=:=) denotes the start of the type of the function name. From
+here any valid type can be written.
+
+Example:
+#+BEGIN_SRC ocaml
+  -- a valid signature and function
+  sig foo : int -> int
+  let foo x = x + 3
+
+
+  -- an example of a dependent signature
+  sig add
+      :  x : nat
+      -> y : int
+      -> if | x > y -> nat
+            | else  -> int
+  let add = (+)
```

Comment:
Why wouldn't this compile? I guess nats aren't a refinement of ints just yet....

(Though I probably ment subtraction as to why it would maybe be an int)

Reply:
sure in that case it might, would be nice, it's a fine example either way